crates/payload/builder/src/traits.rs

```diff
@@ -1,9 +1,58 @@
 //! Trait abstractions used by the payload crate.
 
-use crate::error::PayloadBuilderError;
-use reth_payload_primitives::{BuiltPayload, PayloadBuilderAttributes};
+use crate::{error::PayloadBuilderError, PayloadEvents, PayloadId};
+use reth_payload_primitives::{BuiltPayload, PayloadBuilderAttributes, PayloadTypes};
 use reth_provider::CanonStateNotification;
-use std::future::Future;
+use std::{future::Future, pin::Pin};
+use tokio::sync::oneshot;
+
+pub(crate) type PayloadFuture<P> =
+    Pin<Box<dyn Future<Output = Result<P, PayloadBuilderError>> + Send + Sync>>;
+
+/// A type that can request, subscribe to and resolve payloads.
+#[async_trait::async_trait]
```

Comment:
I'm not too sure about this, are you ok with using the `async_trait` macro or would you rather returning a future?

Reply:
async_trait is fine here, then we could use this as dyn trait as well.